tests/unit/test_pam_tdb.py

```diff
@@ -0,0 +1,323 @@
+import os
+import pam
+import pwd
+import pytest
+import tdb
+
+from collections.abc import Generator
+from contextlib import contextmanager
+from middlewared.utils import crypto
+from middlewared.utils import user_api_key
+from time import monotonic
+
+EXPIRED_TS = 1
+BASE_ID = 1325
+LEGACY_ENTRY_KEY = 'rtpz6u16l42XJJGy5KMJOVfkiQH7CyitaoplXy7TqFTmY7zHqaPXuA1ob07B9bcB'
+LEGACY_ENTRY_HASH = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc05g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+INVALID_HASH_TYPE = '$pbkdf2-canary$29000$CyGktHYOwXgvBYDQOqc05g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+INVALID_SALT = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc0*g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+INVALID_HASH = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc05g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4a*I'
+MISSING_SALT = '$pbkdf2-sha256$29000$$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+MISSING_HASH = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc05g$'
+PAM_DIR = '/etc/pam.d'
+PAM_FILE = 'middleware-api-key'
+PAM_AUTH_LINE = 'auth  [success=1 default=die]      pam_tdb.so debug '
+PAM_FAIL_DELAY = 1.50  # pam_fail_delay is 2s, but we need a little wiggle-room
+
+PAM_FILE_REMAINING_CONTENTS = """
+@include common-auth-unix
+@include common-account
+password        required        pam_deny.so
+session required        pam_deny.so
+"""
+
+
+def write_tdb_file(
+    username: str,
+    hashlist: list[str],
+    expired: bool = False
+) -> int:
+    """
+    Generate a tdb file based on the specified parameters
+    The resulting TDB will have one entry for `username` and
+    a varying amount of hashes.
+
+    Although each hash supports a separate expiry, we are only
+    concerned in these tests in that works overall.
```

Comment:
Awkward sentence.  Looks like it should be 'concerned in these tests in that _it_ works overall.'